man/man1/ccdgen.1
.TH CCDGEN 1NEMO "6 June 2025"

.SH "NAME"
ccdgen \- create or add objects to an image

.SH "SYNOPSIS"
\fBccdgen\fP \fBout=\fP\fIimage-file\fP [parameter=value]

.SH "DESCRIPTION"
\fBccdgen\fP creates or modifies (2D/3D) images by 
adding simple parameterized two/three-dimensional objects. For arbitrary
expressions use \fIccdmath(1NEMO)\fP instead. 
.PP
An added feature is that the objects can be added as if they are
in a "galactic" disk with given inclination and position angle. The
default is a "face-on" view (\fBinc=0\fP). 
No corrections for brightness due to inclination are made however.
.PP
3D image cubes are supported if \fBsize=\fP contains a 3rd non-1 value.

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBout=\fP
Output file, in standard \fIimage(5NEMO)\fP format.
.br
No default.     
.TP
\fBobject=\fP
Object type. Valid are flat, exp, gauss, bar, spiral, noise, test, blobs.
See below for an explanation and their parameters.
.br
Default: flat
.TP
\fBspar=\fP
Parameters for this object. The first parameter is often the peak 
value, and the second one a scale length. Some object have more
complex geometries and require more parameters. The meaning
of the peak value can be modified by setting \fBtotflux=t\fP
(see below). 
.br
For certain objects, the multiple objects can be given, and the sequence
of parameters can be repeated. Currently \fBobject=blobs\fP has this
option for 3D cubes.
.br
Defaults are defined per object, but are typically 1,1... (see below)
.br
The center location of the object is controlled by \fBcenter=\fP (see below),
but defaults to the center of the image.
.TP
\fBsize=\fP
Image size. It is allowed to have a different size in X and Y.
.br
[10,10,1]
.TP
\fBcell=\fP
Cellsize [1]      *not used yet*
.TP
\fBpa=\fP
Position Angle of disk in which the object lives. Counted positive
starting at North through East in the standard astronomical
fashion.
.br
Default: 0
.TP
\fBinc=\fP
Inclination Angle of disk in which object lives. 0 is face on.
.br
Default: 0
.TP
\fBtotflux=t|f\fP
If set to true, the amplitude (always the first parameter in
\fBspar\fP) is interpreted as the total integrated intensity
instead of the peak value of the object.
.br
Default: \fBf\fP.
.TP
\fBfactor=\fP
The factor by which an image is multiplied before the object is added.
Warning: if the amplitude (usually the first of the spar= numbers) is 0,
this will not apply, and the output would equal the input.
.br
Default: 1
.TP
\fBcenter=\fP
Center of object (in 0-based pixel coordinates). The lower left corner
is counted as (0,0) here. This is commonly also known as the reference
pixel (CRPIX in astronomy). 
.br
Default: image center.
.TP
\fBcrpix=\fP
Override/Set crpix
.br
Default: image center
.TP
\fBcrval=\fP
Override/Set crval
.br
(0,0,0) 
.TP
\fBcdelt=\fP
Override/Set cdelt
.br
(1,1,1) 
.TP
\fBseed=\fP
Random seed
.br
[0]    
.TP
\fBheadline=\fP
Random verbiage added to the image.
.br
Default: none
.TP
\fBin=\fP
Input file (optional) to be added to the output file.
.br
Default: none, image is generated.

.SH "OBJECT"
The \fBobject=\fP keyword allows for the following objects, and \fBspar=\fP for their parameters
.nf
.ta +1i +2.5i
OBJECT	EXPR	SPAR
flat	A		A
exp	A.exp(-r/h)	A,h
gauss	A.exp(-r^2/(2h^2))	A,h
bar	A.exp(-r/h)	A,h,e,b   (e=1-minor/major   b=bar position angle w.r.t. disk)
noise	gaussian(m,s)	mean,sigma
spiral	A.exp(-r/h)	A,h,k,p,m,r0,p0
ferrers	A.(1-r^2)^p	A,h,e,b,p (h=size of bar, p=power to 1-r^2)
isothermal	A.(1+r^2)^p	A,h,p
test	x+10y+100z	-
blobs	A,x,y,z,size	they can be repeated for many blobs
\fImaybe to be implemented\fP:

j1x	J1(x)/x
comet	2d proj
jet	jet model w/ power law brightness
shell	2d proj of shell
poisson 
.fi
.PP
A few objects that are in MIRIAD's \fIimgen(1MIRIAD)\fP program and not listed 
here directly, can be simulated or replaced by:
.nf
.ta +1i +2i
level	same as our flat
cluster	same as isothermal, except we also allow \fBp\fP different from -0.5
noise	notice we have a mean, miriad doesn't, also miriad has a crude low approximation
point	use exp with a very small scale length \fBh\fP
gaussian	our gauss, but bmin/bmaj via inc
gauss3	no equivalent, ccdgen does not support all 3D models yet
disk	use ferrers with p=0
.fi
.PP
Most of these objects have a peak or representative intensity value as the first argument,
and a scale length as the second parameter. The next parameters are sometimes scale
free 
.PP
Bars have an additional axis ratio (we mostly use e=1-b/a here) and position angle
of the bar w.r.t. the disk in the disk plane
(see \fIsnaprotate(1NEMO)\fP for an approach if you
know the angle of the bar in the sky plane). A Ferrers bar also needs to know the
power (usuall an integer) to which 1-r^2 is raised.
.PP
Spirals are yet more complex: A and h are the usual peak and exponential scale-length.
k is the wavenumber (related to the pitch angle as tan(pitch)=1/(2.pi.k.r), 
see also \fImkspiral(1NEMO)\fP),
p controls the relative width of the spiral (assumed to be cos^p(m*phi), 
r0 the starting radius of the spiral (defaults to 0), p0 the
phase of the spiral at radius r0, and m the number of spiral arms. In particular,
since the \fIpow(3)\fP function is used internally for non-integer values of \fBp\fP,
the behavior of \fBm\fP for integer and non-integer values of \fBp\fP is different:
for integer values our own internal \fIpowi(3NEMO)\fP is used, and correctly represents.
For example, for m=1 and p=1 you'll get a one armed spiral, but with negative counter-arm.
With p=2 it becomes a 2-armed spiral, as they will get progressively narrower as p remains
even and gets higher. For odd values of p you will again have a positive and negative
spiral, and as p gets larger, the arms get narrower.   By making sure \fBp\fP becomes
non-integer, e.g. p=2.0001, the negative arm becomes 0.

.SH "EXAMPLES"
Make a exponential bar:
.EX
 ccdgen bar.ccd   object=bar spar=1,10,0.5,30 size=128,128
 ccdplot bar.ccd
.EE

Make a simple barred galaxy by adding an exponential disk and
a (smaller) exponential bar:
.EX

 ccdgen - object=exp spar=1,40 pa=60 inc=45 size=256 |\\
 ccdgen in=- out=bar.ccd object=bar spar=10,6,0.5,70 pa=60 inc=45

.EE
or if you want them to be a bit more astronomical, you'll need to make the units
come out in degrees for the conversion to FITS, viz.:
.EX

 d=3600
 ccdgen - object=exp spar=1,40/$d pa=60 inc=45 size=256 cdelt=-1/$d,1/$d |\\
 ccdgen - object=bar spar=10,6/$d,0.5,70 pa=60 inc=45 in=- |\\
 ccdfits - bar.fits radecvel=t

.EE
and the option to fill your cube with some random blobs and noise is a bit more involved.
We create a simple text file \fBb100.tab\fP containing the random parameters of the blobs
(Amplitude, x, y, z, size):
.EX

nemoinp 1:100 |\\
  tabmath - - 'ranu(0,1),ranu(10,100),ranu(10,100),ranu(10,100),ranu(0,5)' all\\
              > b100.tab

ccdgen - blobs @b100.tab size=128,128,128 |\\
  ccdgen - noise 0,0.02 in=- |\\
  ccdsmooth - - 5   |\\
  ccdsky - -  |\\
  ccdfits - junk2.fits radecvel=t ndim=4

.EE
Making a new data section (in this case just noise)
by stealing the header from existing \fIfits(5NEMO)\fP file. 
In \fIbash\fP notation:
.EX

 fitshead NGC5194_91112.fits > tmp.fitshead
 nx=$(grep NAXIS1 tmp.fitshead|awk '{print $3}')
 ny=$(grep NAXIS2 tmp.fitshead|awk '{print $3}')
 nz=$(grep NAXIS3 tmp.fitshead|awk '{print $3}')
 ccdgen - object=noise spar=0,1 size=$nx,$ny,$nz |\\
     ccdfits - noise.fits fitshead=NGC5194_91112.fits
  
.EE
A test cube234.fits with value x + 10y + 100z can be made as follows:
.EX
 ccdmath out=- fie=%x+10*%y+100*%z size=2,3,4 | ccdfits - cube234.fits
.EE

.SH "CAVEATS"
Seting the WCS is not working well. There is cell=, there is crpix,cdelt,crval


.SH "SEE ALSO"
ccdmath(1NEMO), snaprotate(1NEMO), image(5NEMO)

.PP
https://github.com/GalSim-developers/GalSim
.PP
imfit::makeimage
.PP
There are also good simulators in starlab  and amuse

.SH "FILES"
src/image/trans

.SH "AUTHOR"
Peter Teuben

.SH "UPDATE HISTORY"
.nf
.ta +1.5i +5.5i
4-Jan-05	V0.1 Created	PJT
6-jan-05	V0.7 added (many features and) factor=	PJT
8-jan-05	V0.8 add #arms parameter to spiral	PJT
31-jan-12	V0.9 added object=test	PJT
12-mar-2020	V1.0 added object=blobs		PJT
8-may-2022	V1.2 moved center= down the arg list 	PJT
8-dec-2022	V2.0 moved in= down the arg list	PJT
.fi


src/image/trans/ccdgen.c
/* 
 * CCDGEN:   create 2D 'astronomical' type objects, modeled after MIRIAD'd imgen
 *
 *      4-jan-05        V0.1: for Kartik's bars modeling experiment 
 *                      (IMGEN is just tooooo cumbersome, and so is ccdmath) 
 *      6-jan-05        V0.7: added many more models and features. added factor=
 *      8-jan-05        V0.8: add m!=2 multi-arm spirals
 *      5-aug-11        V0.9: object = test
 *     15-oct-2014      V0.10:  object=noise now can do 3D
 *     12-mar-2020      V1.0    object=blobs in 3D
 *        dec-2022      V2.0:  moved in= down the argument list
 * TODO:
 *    - find out why the normalization was PI, and not TWO_PI, which worked before.
 *       (this happened when I changed from 1 to 1/3600 scaling factor in the examples)
 *    - add a few more of the imgen models
 *    - replace crpix/crval/cdelt with cell= and maybe radec= ? as in imgen
 *    - if A=0, then factor most like won't work (hard to fix the way it's written)
 *
 */

#include <stdinc.h>
#include <ctype.h>
#include <getparam.h>
#include <vectmath.h>
#include <filestruct.h>
#include <history.h>
#include <strlib.h>
#include <image.h>

string defv[] = {
  "out=???\n       Output file",
  "object=flat\n   Object type (test,flat,exp,gauss,bar,spiral,....)",
  "spar=\n         Parameters for this object",
  "size=10,10,1\n  2- or 3D dimensions of map/cube (only if no input file given)",
  "cell=1\n        Cellsize",
  "pa=0\n          Position Angle of disk in which object lives",
  "inc=0\n         Inclination Angle of disk in which object lives",
  "totflux=f\n     Interpret peak values (spar[0]) as total flux instead",
  "factor=1\n      Factor to multiple input image by before adding object",
  "center=\n       Center of object (defaults to map(reference)center) in pixels 0..size-1",
  "crpix=\n        Override/Set crpix (1,1,1) // ignored ** mapcenter if left blank **",
  "crval=\n        Override/Set crval (0,0,0) // ignored",
  "cdelt=\n        Override/Set cdelt (1,1,1) // ignored",
  "seed=0\n        Random seed",
  "headline=\n     Random veriage for the history",
  "in=\n           Input file (optional) to be added to the new file",
  "VERSION=2.0\n   8-dec-2022 PJT",
  NULL,
};

string usage = "image creation/modification with objects";

#ifndef HUGE
# define HUGE 1.0e20
#endif

#define MAXIMAGE 20
#define MAXPAR  32768           /* max in nemoinp */

imageptr iptr;	                /* pointers to (input) image */

#define MAXNAX 3

double crval[MAXNAX], crpix[MAXNAX], cdelt[MAXNAX];
int nwcs = 0;

real spar[MAXPAR];
int npar;

real pa,inc,center[3];
real sinp, cosp, sini, cosi;

bool Qtotflux;
real factor;
real surface=1.0;


local void do_create(int nx, int ny, int nz);

local string objects[] = {
  "flat", "exp", "gauss", "bar", "ferrers",
  "spiral", "noise", "jet", "j1x", "isothermal",
  "comet", "shell", "point", "test", "blobs",
  NULL,
};
  
local void object_test(int npars, real *pars);
local void object_flat(int npars, real *pars);
local void object_exp(int npars, real *pars);
local void object_gauss(int npars, real *pars);
local void object_bar(int npars, real *pars);
local void object_ferrers(int npars, real *pars);
local void object_spiral(int npars, real *pars);
local void object_noise(int npars, real *pars);
local void object_j1x(int npars, real *pars);
local void object_isothermal(int npars, real *pars);
local void object_comet(int npars, real *pars);
local void object_jet(int npars, real *pars);
local void object_shell(int npars, real *pars);
local void object_point(int npars, real *pars);
local void object_blobs(int ndim, int npars, real *pars);


void nemo_main(void)
{
  string  fnames;
  stream  instr;                      /* input file (optional) */
  stream  outstr;                     /* output file */
  int     size[3],nx, ny, nz;         /* size of scratch map */
  int     ncen;
  string  object;
  string  headline;
  int seed = init_xrandom(getparam("seed"));
  int ndim = 0;

  object = getparam("object");
  npar = nemoinpr(getparam("spar"),spar,MAXPAR);
  if (npar < 0) error("Syntax error %s",getparam("spar"));

  Qtotflux = getbparam("totflux");
  factor = getdparam("factor");

  pa = getdparam("pa");
  inc = getdparam("inc");
  sinp = sin(pa*PI/180.0);
  cosp = cos(pa*PI/180.0);
  sini = sin(inc*PI/180.0);
  cosi = cos(inc*PI/180.0);
  dprintf(0,"%s: disk with pa=%g inc=%g\n",object,pa,inc);
  dprintf(1,"pa(%g %g) inc(%g %g)\n",sinp,cosp,sini,cosi);

  ncen = nemoinpr(getparam("center"),center,2);
  if (ncen<0) error("Syntax error %s",getparam("center"));
  if (ncen==1) center[1] = center[0];

  nwcs = nemorinpd(getparam("crval"),crval,MAXNAX,0.0,FALSE);
  nwcs = nemorinpd(getparam("cdelt"),cdelt,MAXNAX,1.0,FALSE);
  nwcs = nemorinpd(getparam("crpix"),crpix,MAXNAX,1.0,FALSE);

  outstr = stropen (getparam("out"),"w");  /* open output file first ... */
  
  if (hasvalue("in")) {
    instr = stropen(getparam("in"),"r");    /* open file */
    read_image (instr, &iptr);
  } else {
    dprintf(0,"Generating a map from scratch\n");
    switch (nemoinpi(getparam("size"),size,3)) {
    case 1:			/*  nx[,nx,1] */
      size[1] = size[0];
      size[2] = 1;
      ndim = 2;
      break;
    case 2:			/*  nx,ny[,1] */
      size[2] = 1;
      ndim = 2;
      break;
    case 3:			/*  nx,ny,nz  */
      ndim = 3;
      break;
    case 0:			/*  [10,10,1] */
      dprintf(0,"Cannot have no size, default 10 assumed\n");
      size[0] = size[1] = 10;
      size[2] = 1;
      ndim = 2;
      break;
    default:			/* --- some error --- */
      error("Syntax error in size keyword\n");
    }
    nx = size[0];
    ny = size[1];
    nz = size[2];
    do_create(nx,ny,nz);
#if 0
    if (nwcs == 0) {    /* last one for crpix */
      warning("going to set the center of the map in new NEMO convention");
      crpix[0] = (nx-1)/2.0;
      crpix[1] = (ny-1)/2.0;
      crpix[2] = (nz-1)/2.0;
    }
#endif
  }
  dprintf(1,"Shape: %d x %d x %d\n", size[0],size[1],size[2]);


  if (!Object(iptr))
    Object(iptr) = scopy(object);
  
  if (ncen==0) {   /* fix center if it has not been set yet */
    center[0] = (Nx(iptr)-1)/2.0;     /* 0 based center= */
    center[1] = (Ny(iptr)-1)/2.0;
  }
  Xref(iptr) = center[0];
  Yref(iptr) = center[1];
  Zref(iptr) = 0.0;
  Axis(iptr) = 1;
  dprintf(0,"%s: center pixel: %g %g\n",object,center[0],center[1]);
  surface = Dx(iptr)*Dy(iptr);
  surface = ABS(surface);



  if (streq(object,"flat"))
    object_flat(npar,spar);
  else if (streq(object,"exp"))
    object_exp(npar,spar);
  else if (streq(object,"gauss"))
    object_gauss(npar,spar);
  else if (streq(object,"bar"))
    object_bar(npar,spar);
  else if (streq(object,"ferrers"))
    object_ferrers(npar,spar);
  else if (streq(object,"spiral"))
    object_spiral(npar,spar);
  else if (streq(object,"noise"))
    object_noise(npar,spar);
  else if (streq(object,"jet"))
    object_jet(npar,spar);
  else if (streq(object,"j1x"))
    object_j1x(npar,spar);
  else if (streq(object,"isothermal"))
    object_isothermal(npar,spar);
  else if (streq(object,"comet"))
    object_comet(npar,spar);
  else if (streq(object,"shell"))
    object_shell(npar,spar);
  else if (streq(object,"point"))
    object_point(npar,spar);
  else if (streq(object,"test"))
    object_test(npar,spar);
  else if (streq(object,"blobs"))
    object_blobs(ndim,npar,spar);
  else
    error("Unknown object %s",object);

  minmax_image(iptr);
  
  if (hasvalue("headline"))
    set_headline(getparam("headline"));
  write_image (outstr,iptr);         /* write image to file */
  strclose(outstr);
}


/*
 *  create new map from scratch, using %x and %y as position parameters 
 *		0..nx-1 and 0..ny-1
 */
local void do_create(int nx, int ny, int nz)
{
    double m_min, m_max, total;
    real   fin[5], fout;
    int    ix, iy, iz;
    int    badvalues;
    
    m_min = HUGE; m_max = -HUGE;
    total = 0.0;		/* count total intensity in new map */
    badvalues = 0;		/* count number of bad operations */

    if (nz > 0) {
      // warning("cube");
      if (!create_cube (&iptr, nx, ny, nz))	/* create default empty image */
        error("Could not create 3D image from scratch");
#if 0      
      Axis(iptr) = 1;      /* set linear axistype with a fits-style reference pixel */
#endif
      wcs_f2i(3,crpix,crval,cdelt,iptr);

      for (iz=0; iz<nz; iz++) {
        fin[2] = iz-crpix[2]+1;   /* crpix is 1 for first pixel (FITS convention) */
        for (iy=0; iy<ny; iy++) {
	  fin[1] = iy-crpix[1]+1;
	  for (ix=0; ix<nx; ix++) {
	    fout = 0.0;
	    CubeValue(iptr,ix,iy,iz) = fout;
	    m_min = MIN(m_min,fout);         /* and check for new minmax */
	    m_max = MAX(m_max,fout);
	    total += fout;                   /* add up totals */
	  }
        }
      }
    } else {
      warning("2d-map");
      if (!create_image (&iptr, nx, ny))	
        error("Could not create 2D image from scratch");
      wcs_f2i(2,crpix,crval,cdelt,iptr);

      for (iy=0; iy<ny; iy++) {
	fin[1] = iy;
	for (ix=0; ix<nx; ix++) {
	  fout = 0.0;
	  MapValue(iptr,ix,iy) = fout;
	  m_min = MIN(m_min,fout);         /* and check for new minmax */
	  m_max = MAX(m_max,fout);
	  total += fout;                   /* add up totals */
	}
      }
    } 
    
    MapMin(iptr) = m_min;
    MapMax(iptr) = m_max;

    dprintf(1,"New min and max in map are: %f %f\n",m_min,m_max);
    dprintf(1,"New total brightness/mass is %f\n",
			total*Dx(iptr)*Dy(iptr));
    if (badvalues)
    	warning ("There were %d bad operations in dofie",badvalues);
}


local void object_test(int npars, real *pars)
{
  int i,j,k;
  int nx = Nx(iptr);
  int ny = Ny(iptr);
  int nz = Ny(iptr);
  real A = 1.0;

  for (k=0; k<nz; k++)
    for (j=0; j<ny; j++)
      for (i=0; i<nx; i++)
	CubeValue(iptr,i,j,k) = i + 10*j + 100*k;
}


local void object_flat(int npars, real *pars)
{
  int i,j;
  int nx = Nx(iptr);
  int ny = Ny(iptr);
  real A = 1.0;

  if (A==0) return;

  if (npar > 0) A = pars[0];

  if (Qtotflux) A /= (nx*ny);

  for (j=0; j<ny; j++)
    for (i=0; i<nx; i++)
      MapValue(iptr,i,j) = factor*MapValue(iptr,i,j) + A;

}

local void object_exp(int npars, real *pars)
{
  int i,nx = Nx(iptr);
  int j,ny = Ny(iptr);
  real A = 1.0;
  real h = 1.0;
  real x1,y1,x2,y2,r,arg,value;

  if (npar > 0) A = pars[0];
  if (npar > 1) h = pars[1];
  dprintf(0,"exp: %g %g\n",A,h);

  if (A==0) return;

  if (Qtotflux) {
    A /= (PI*h*h/surface);
    dprintf(0,"exp: A->%g\n",A);
  }

  for (j=0; j<ny; j++) {
    y1 = (j-center[1])*Dy(iptr) + Ymin(iptr);
    for (i=0; i<nx; i++) {
      x1 = (i-center[0])*Dx(iptr) + Xmin(iptr);
      x2 =  -x1*sinp - y1*cosp;
      y2 =   x1*cosp - y1*sinp;
      r = sqrt(x2*x2 + sqr(y2/cosi));
      arg = r/h;
      dprintf(2,"%d %d : %g %g %g %g  %g\n",i,j,x1,y1,x2,y2,arg);
      value = (arg < 80) ? A * exp(-arg) : 0.0;
      MapValue(iptr,i,j) = factor*MapValue(iptr,i,j) + value;
    }
  }
}

local void object_gauss(int npars, real *pars)
{
  int i,nx = Nx(iptr);
  int j,ny = Ny(iptr);
  real A = 1.0;
  real h = 1.0;
  real x1,y1,x2,y2,r,arg,value;

  if (npar > 0) A = pars[0];
  if (npar > 1) h = pars[1];
  dprintf(0,"gauss: %g %g\n",A,h);

  if (A==0) return;

  if (Qtotflux) {
    A /= (PI*h*h/surface);
    dprintf(0,"gauss: A->%g\n",A);
  }

  for (j=0; j<ny; j++) {
    y1 = (j-center[1])*Dy(iptr) + Ymin(iptr);
    for (i=0; i<nx; i++) {
      x1 = (i-center[0])*Dx(iptr) + Xmin(iptr);
      x2 =  -x1*sinp - y1*cosp;
      y2 =   x1*cosp - y1*sinp;
      r = sqrt(x2*x2 + sqr(y2/cosi));
      arg = r/h;
      dprintf(2,"%d %d : %g %g %g %g   %g\n",i,j,x1,y1,x2,y2,arg);
      value = (arg < 13) ?  A * exp(-0.5*arg*arg) : 0.0;
      MapValue(iptr,i,j)  = factor*MapValue(iptr,i,j) + value;
    }
  }
}

local void object_bar(int npars, real *pars)
{
  int i,nx = Nx(iptr);
  int j,ny = Ny(iptr);
  real A = 1.0;
  real h = 1.0;
  real e = 0.0;
  real b = 0.0;
  real x1,y1,x2,y2,x3,y3,r,arg,value;
  real sinb,cosb,omba;

  if (npar > 0) A = pars[0];
  if (npar > 1) h = pars[1];
  if (npar > 2) e = pars[2];
  if (npar > 3) b = pars[3];
  dprintf(0,"bar: %g %g %g %g\n",A,h,e,b);

  sinb = sin(b*PI/180.0);
  cosb = cos(b*PI/180.0);

  if (A==0) return;

  if (Qtotflux) {
    A /= (PI*h*h*(1-e)/surface);
    dprintf(0,"bar: A->%g\n",A);
  }
  dprintf(1,"bar b=%g\n",b);

  for (j=0; j<ny; j++) {
    y1 = (j-center[1])*Dy(iptr) + Ymin(iptr);
    for (i=0; i<nx; i++) {
      x1 = (i-center[0])*Dx(iptr) + Xmin(iptr);

      x2 =  -x1*sinp - y1*cosp;
      y2 =   (x1*cosp - y1*sinp)/cosi;

      x3 =   x2*cosb - y2*sinb;
      y3 = (x2*sinb  + y2*cosb)/(1-e);

      r = sqrt(x3*x3+y3*y3);
      arg = r/h;
      value = (arg < 100) ? A * exp(-arg) : 0.0;
      MapValue(iptr,i,j) = factor*MapValue(iptr,i,j) + value;
    }
  }
}


local void object_ferrers(int npars, real *pars)
{
  int i,nx = Nx(iptr);
  int j,ny = Ny(iptr);
  int l, lmax;
  real A = 1.0;   /* peak */
  real h = 1.0;   /* size */
  real e = 0.0;   /* 1-b/a */
  real b = 0.0;   /* phi */
  real p = 1.0;   /* power factor */
  real x1,y1,x2,y2,x3,y3,r,arg,value;
  real amp,phi,sum;
  real sinb,cosb;

  if (npar > 0) A = pars[0];
  if (npar > 1) h = pars[1];
  if (npar > 2) e = pars[2];
  if (npar > 3) b = pars[3];
  if (npar > 4) p = pars[4];
  dprintf(0,"ferrers: %g %g %g %g %g\n",A,h,e,b,p);

  if (A==0) return;

  lmax = Qtotflux ? 2 : 1;

  sinb = sin(b*PI/180.0);
  cosb = cos(b*PI/180.0);

  for (l=0; l<lmax; l++) {     /* 1st loop: sum up the flux   if in 2nd loop: normalize */
    if (l==0) 
      sum = 0.0;
    else {
      A /= sum;
      dprintf(0,"ferrers: A->%g\n",A);
    } 
    for (j=0; j<ny; j++) {
      y1 = (j-center[1])*Dy(iptr) + Ymin(iptr);
      for (i=0; i<nx; i++) {
	x1 = (i-center[0])*Dx(iptr) + Xmin(iptr);
	
	x2 =  -x1*sinp - y1*cosp;
	y2 =  (x1*cosp - y1*sinp)/cosi;

	x3 =   x2*cosb - y2*sinb;
	y3 = (x2*sinb  + y2*cosb)/(1-e);

	r = (x3*x3+y3*y3)/(h*h);
	value = r < 1 ? A * pow(1.0-r, p) : 0.0;

	if (Qtotflux) {   
	  if (l==0) 
	    sum += value;
	  else
	    MapValue(iptr,i,j) = factor*MapValue(iptr,i,j) + value;
	} else
	  MapValue(iptr,i,j) = factor*MapValue(iptr,i,j) + value;
      } /* i */
    } /* j */
  } /* k */
}


local void object_spiral(int npars, real *pars)
{
  int i,nx = Nx(iptr);
  int j,ny = Ny(iptr);
  int l, lmax;
  real A = 1.0;   /* Amplitude */
  real h = 1.0;   /* radial scale length */
  real k = 1.0;   /* wave number */
  real p = 1.0;   /* 1/2 power of cos */
  real r0 = 0.0;  /* starting radius */
  real p0 = 0.0;  /* starting angle */
  int m = 2;      /* number of arms */
  real x1,y1,x2,y2,x3,y3,r,arg,value;
  real amp,phi,sum;
  bool Qint, pint;

  if (npar > 0) A = pars[0];
  if (npar > 1) h = pars[1];
  if (npar > 2) k = pars[2];
  if (npar > 3) p = pars[3];
  if (npar > 4) m =  (int) pars[4];  /* note rounding ? */
  if (npar > 5) r0 = pars[5];
  if (npar > 6) p0 = pars[6];

  dprintf(0,"spiral: %g %g   %g %g %d   %g %g\n",A,h,k,p,m,r0,p0);

  if (A==0) return;

  p0 *= PI/180.0;  /* convert from degrees to radians */
  k *= TWO_PI;     /* angles are 2.PI.k.r , so absorb 2.PI.k in one */
  pint = (int) p;
  Qint = (p-pint == 0);
  if (Qint)
    warning("Integer power p = %d\n",pint);

  lmax = Qtotflux ? 2 : 1;

  for (l=0; l<lmax; l++) {     /* 1st loop: sum up the flux   if in 2nd loop: normalize */
    if (l==0) 
      sum = 0.0;
    else {
      A /= sum;
      dprintf(0,"spiral: A->%g\n",A);
    } 
    for (j=0; j<ny; j++) {
      y1 = (j-center[1])*Dy(iptr) + Ymin(iptr);
      for (i=0; i<nx; i++) {
	x1 = (i-center[0])*Dx(iptr) + Xmin(iptr);
	
	x2 =  -x1*sinp - y1*cosp;
	y2 =  (x1*cosp - y1*sinp)/cosi;
	r = sqrt(x2*x2+y2*y2);
	if (r < r0) 
	  value = 0.0;
	else {
	  /* ? should match this up better so we can connect bar and spiral ? */
	  phi = atan2(y2,x2) + k*(r-r0) + p0;
	  if (Qint) {
	    amp = powi(cos(m*phi),pint);   /* these can come out negative for odd p's !! */
	  } else
	    amp = powd(cos(m*phi),(double)p);
	  arg = r/h;
	  value = (arg < 80) ? A * amp * exp(-arg) :  0.0;
	}
	if (Qtotflux) {   
	  if (l==0) 
	    sum += value;
	  else
	    MapValue(iptr,i,j) = factor*MapValue(iptr,i,j) + value;
	} else
	  MapValue(iptr,i,j) = factor*MapValue(iptr,i,j) + value;
      } /* i */
    } /* j */
  } /* k */
}

local void object_noise(int npars, real *pars)
{
  int i,j,k;
  int nx = Nx(iptr);
  int ny = Ny(iptr);
  int nz = Nz(iptr);
  double m = 1.0;
  double s = 1.0;

  if (npar > 0) m = pars[0];
  if (npar > 1) s = pars[1];

  dprintf(0,"noise:%g %g\n",m,s);

  if (s==0) return;

  if (Qtotflux) {
    m /= (nx*ny);
    s /= (nx*ny);
    dprintf(0,"noise: m->%g  s->%g\n",m,s);
  }
  
  for (k=0; k<nz; k++)
    for (j=0; j<ny; j++)
      for (i=0; i<nx; i++)
	CubeValue(iptr,i,j,k) = factor*CubeValue(iptr,i,j,k) + grandom(m,s);
}



local void object_j1x(int npars, real *pars)
{
  warning("alas, j1x not supported yet");
}

local void object_isothermal(int npars, real *pars)
{
  int i,nx = Nx(iptr);
  int j,ny = Ny(iptr);
  int l, lmax;
  real A = 1.0;   /* peak */
  real h = 1.0;   /* size */
  real p = -0.5;  /* power factor */
  real x1,y1,x2,y2,x3,y3,r,arg,value;
  real amp,phi,sum;
  real sinb,cosb;

  if (npar > 0) A = pars[0];
  if (npar > 1) h = pars[1];
  if (npar > 2) p = pars[2];
  dprintf(0,"isothermal: %g %g %g \n",A,h,p);

  if (A==0) return;

  lmax = Qtotflux ? 2 : 1;

  for (l=0; l<lmax; l++) {     /* 1st loop: sum up the flux   if in 2nd loop: normalize */
    if (l==0) 
      sum = 0.0;
    else {
      A /= sum;
      dprintf(0,"isothermal: A->%g\n",A);
    } 
    for (j=0; j<ny; j++) {
      y1 = (j-center[1])*Dy(iptr) + Ymin(iptr);
      for (i=0; i<nx; i++) {
	x1 = (i-center[0])*Dx(iptr) + Xmin(iptr);
	
	x2 =  -x1*sinp - y1*cosp;
	y2 =  (x1*cosp - y1*sinp)/cosi;

	r = (x2*x2+y2*y2)/(h*h);
	value = A * pow(1.0+r, p);

	if (Qtotflux) {   
	  if (l==0) 
	    sum += value;
	  else
	    MapValue(iptr,i,j) = factor*MapValue(iptr,i,j) + value;
	} else
	  MapValue(iptr,i,j) = factor*MapValue(iptr,i,j) + value;
      } /* i */
    } /* j */
  } /* k */
}

local void object_comet(int npars, real *pars)
{
  warning("alas, comet not supported yet");
}

local void object_jet(int npars, real *pars)
{
  warning("alas, jet not supported yet");
}

local void object_shell(int npars, real *pars)
{
  warning("alas, shell not supported yet");
}

local void object_point(int npars, real *pars)
{
  int nx = Nx(iptr);
  int ny = Ny(iptr);
  int i = nx/2;
  int j = ny/2;

  if (npars == 0) error("object=point requires a value");

  MapValue(iptr,i,j) = factor*MapValue(iptr,i,j) + pars[0];
}


local void object_blobs(int ndim, int npars, real *pars)
{
  int x0, ix, nx = Nx(iptr);
  int y0, iy, ny = Ny(iptr);
  int z0, iz, nz = Ny(iptr);
  int i, j, k;
  int npar = 5;  // peak, x, y, z, s
  int ns, ib, nblobs = npars/npar;
  real pk, s0, dx, dy, dz, arg;

  if (npars == 0) error("object=blobs requires a value");
  if (npars % npar != 0) error("object=blobs not commensurate to npar=%d",npar);
  if (ndim != 3) error("object=blobs not yet coded for 2d");

  dprintf(1,"Found %d blobs\n",nblobs);

  for (ib=0; ib<nblobs; ib++) {
    pk =      pars[ib*npar + 0];
    x0 = (int)pars[ib*npar + 1];
    y0 = (int)pars[ib*npar + 2];
    z0 = (int)pars[ib*npar + 3];
    s0 =      pars[ib*npar + 4];
    ns = (int) (5*s0);
    dprintf(1,"%d  %g  %d %d %d %g\n",ib,pk,x0,y0,z0,s0);

    for (ix=x0-ns; ix<=x0+ns; ix++) {
      if (ix<0 || ix>=nx) continue;
      dx = ix-x0;
      for (iy=y0-ns; iy<=y0+ns; iy++) {
	if (iy<0 || iy>=ny) continue;
	dy = iy-y0;
	for (iz=z0-ns; iz<=z0+ns; iz++) {
	  if (iz<0 || iz>=nz) continue;	  
	  dz = iz-z0;
	  arg = (dx*dx + dy*dy + dz*dz)/(s0*s0);
	  CubeValue(iptr,ix,iy,iz) += pk * exp(-arg);
	}//iz
      }//iy
    }//iz
  }//ib
}
